### DISCUSSION TITLE
How to investigate global routing problems

### DISCUSSION THREAD

**Discussion by oharboe** [The-OpenROAD-Project/OpenROAD-flow-scripts#1338] [General]
I seem to recall there was a feature that included the nets that were involved in the congestion under the DRC?

I was testing out #1333 on sky130hd/microwatt because it is the only design that has hard global routing in the ORFS repository.

mpl2 doesn't solve the global routing issue in sky130hd/microwatt, not that I know if it can...

```
diff --git a/flow/designs/sky130hd/microwatt/config.mk b/flow/designs/sky130hd/microwatt/config.mk
index 4b4eb79d..33d5812d 100644
--- a/flow/designs/sky130hd/microwatt/config.mk
+++ b/flow/designs/sky130hd/microwatt/config.mk
@@ -31,4 +31,5 @@ export CTS_CLUSTER_SIZE = 30
 
 export export SETUP_SLACK_MARGIN = 0.2
 
-export GLOBAL_ROUTE_ARGS=-congestion_iterations 100 -verbose
+export SYNTH_HIERARCHICAL = 1
+export RTLMP_FLOW = True
```

global routing DRC report after 20 iterations:

![image](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/2798822/88af22df-2524-4c19-834e-b84c8f036641)

**Comment by vijayank88** [comment]
@oharboe 
With the hierarchical synthesis have you checked its placement/routing heat map?
Because it's implemented for flat design with fixed die size.

**Reply by oharboe** [reply]
Placement heatmap:

![image](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/2798822/638f77f4-d586-4feb-a9d7-e48f7ee91cc6)

Routing congestion heatmap:

![image](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/2798822/6188bba5-2aaa-4095-8dc9-1b73b71bec53)

**Comment by oharboe** [comment]
> @oharboe With the hierarchical synthesis have you checked its placement/routing heat map?

Above.

> Because it's implemented for flat design with fixed die size.

Why is a fixed die-size required?

**Reply by vijayank88** [reply]
@oharboe 
Microwatt design got fabricated with sky130 PDK. So it has fixed die size of 2920x3520.

**Reply by oharboe** [reply]
Yes, but is a fixed die size required for nets to be included in the global route DRC report?

**Reply by oharboe** [reply]
I just checked with a flattened synthesis and without MPL2 (i.e. how it is in master) and there are no nets listed in the global route DRC report.

Should there have been?

![image](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/2798822/c4dbd528-8571-4c6d-8332-947c805888b1)

![image](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/2798822/c9d720fc-7074-407f-a72b-3f4e192c985a)

**Reply by vijayank88** [reply]
With default `config.mk` the design passes all stages completely.
Do you want to update following configuration and check the routing?
```
 export export SETUP_SLACK_MARGIN = 0.2
 
-export GLOBAL_ROUTE_ARGS=-congestion_iterations 100 -verbose
+export SYNTH_HIERARCHICAL = 1
+export RTLMP_FLOW = True
```

**Reply by oharboe** [reply]
The design passes here as well. I am using this design to study progress reporting on hard global routing: https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/pull/1333

**Reply by vijayank88** [reply]
Possible can you provide `make global_route_issue` test case?
Let me reproduce the screenshot you've attached, for better understanding.

**Reply by oharboe** [reply]
To reproduce:

1. Use https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/pull/1333
2. Comment out GLOBAL_ROUTE_ARGS, see below:
3. Run `make DESIGN_CONFIG=designs/sky130hd/microwatt/config.mk results/sky130hd/microwatt/base/5_1_grt.odb`
4. *While* global route is running, start a new shell, run `ls reports/sky130hd/microwatt/base/congestion*.rpt`. When you see .rpt files appearing, run `make DESIGN_CONFIG=designs/sky130hd/microwatt/config.mk gui_cts` and load for instance `reports/sky130hd/microwatt/base/congestion-10.rpt`
5. Look at DRC errors, there's no information about which nets are congested.


```
$ git diff
diff --git a/flow/designs/sky130hd/microwatt/config.mk b/flow/designs/sky130hd/microwatt/config.mk
index 4b4eb79d..90745a5d 100644
--- a/flow/designs/sky130hd/microwatt/config.mk
+++ b/flow/designs/sky130hd/microwatt/config.mk
@@ -31,4 +31,4 @@ export CTS_CLUSTER_SIZE = 30
 
 export export SETUP_SLACK_MARGIN = 0.2
 
-export GLOBAL_ROUTE_ARGS=-congestion_iterations 100 -verbose
+#export GLOBAL_ROUTE_ARGS=-congestion_iterations 100 -verbose
```

**Reply by vijayank88** [reply]
Yes, I understand your requirement. It's similar to my request already raised: https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/issues/1216

Because during congestion, flow get terminated without writing current stage .odb file.
Even though we have `congestion.rpt` it can't be located with OpenROAD gui.
Only if we run with test case, we can able to trace those nets with x marker in GUI.

To view congestion report we need global routing stage database not cts stage `.odb`.
Hope I've cleared your issue on why congestion marker not show as per the steps you followed.

**Reply by oharboe** [reply]
Actually, the net information should be saved in the .rpt file, the .odb file doesn't enter into it, as I understand.

From https://github.com/The-OpenROAD-Project/OpenROAD/blob/36bd481ac8a9edf401ba5e5b6bb0c379ac958467/src/grt/src/fastroute/src/utility.cpp#L2367
, that's global routing, we see that it saves net information, if there is any.

This leads to the next question: why isn't there any net information in this case?

**Reply by oharboe** [reply]
@maliberty Should there be any net information in this case, looks like there should have been reading the source code?

**Reply by maliberty** [reply]
There should be so open an issue if needed

**Reply by oharboe** [reply]
Done: https://github.com/The-OpenROAD-Project/OpenROAD/issues/3867
